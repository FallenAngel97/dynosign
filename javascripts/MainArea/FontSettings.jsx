import React from 'react'
import PropTypes from 'prop-types';
import Select from 'react-select';
import { ipcRenderer } from 'electron'

export default class FontSettings extends React.Component {
  constructor (props) {
    super(props);
    const fonts = ipcRenderer.sendSync('getfonts', 'ping');
    this.state = {
      defaultFont: { defaultFont: { value: fonts[0], label: fonts[0].family } }
    }
    this.options = fonts.map((font) => {
      return { value: font, label: font.family };
    });
    this.fontChange = this.fontChange.bind(this)
  }
  fontChange (font) {
    this.setState({ defaultFont: font })
  }
  render () {
    return (
      <>
        <div id='fontSelector'>
          <Select onChange={this.fontChange} options={this.options} value={this.state.defaultFont}/>
        </div>
        <input id='fontSize_picker' defaultValue={10} />
      </>
    )
  }
}

FontSettings.propTypes = {
  options: PropTypes.array,
  defaultFont: PropTypes.object
}
