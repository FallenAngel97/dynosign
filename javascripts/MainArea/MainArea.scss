#mainArea {
    height: calc(100% - (var(--height-of-top-bar) + 5px) * 2);
    grid-column: 2;
}
.drawAnimateSelector {
    background: linear-gradient(to top, #232323, #6d6d6d);
    box-shadow: 1px 1px 2px 0px rgba(0,0,0,.2);
    border:none;
    border-radius: 5px;
    padding: 5px 10px;
    font-family: 'FranklinGothicDemiC';
    color: #fff;
    font-size: 16px;
    margin:2px;
    cursor: pointer;
    transition: .2s ease-in-out box-shadow;
    &:hover {
        box-shadow: 1px 1px 4px 2px rgba(0,0,0,.2);
    }
}
#drawingArea {
    margin-top:10px;
    background: #fff;
    height: calc(100% - 50px);
    box-shadow: 0 0 5px 2px rgba(0,0,0,.2);
}
.canvaslayer {
    position: absolute;
}
#fontSelector {
    display: inline-block;
    width: 187px;
}