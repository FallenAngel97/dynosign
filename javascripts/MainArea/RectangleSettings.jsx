import React from 'react';
import './RectangleSettings.scss';

export default class RectangleSettings extends React.Component {
  render () {
    return (
      <div id='rectangle_settings'>
        Left: <input />
        Top: <input />
        Height: <input />
        Width: <input />
      </div>
    )
  }
}
