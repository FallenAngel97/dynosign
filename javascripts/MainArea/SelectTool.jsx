import React from "react";
import {connect} from "react-redux";

class SelectTool extends React.Component {
    constructor(props) {
        super(props);
        this.prevX = 0;
        this.currX = 0;
        this.prevX = 0;
        this.prevY = 0;
        this.flag = false;
        this.findxy = this.findxy.bind(this);
        this.draw = this.draw.bind(this);
        this.timerMarchingAnts = undefined;
    }
    draw(offsetDash) {
        var ctx = this.canvas.getContext('2d');
        ctx.beginPath();
        ctx.clearRect(0, 0, this.canvas.width, this.canvas.height);
        ctx.setLineDash([6]);
        ctx.lineDashOffset = offsetDash;
        const diffX = this.currX - this.prevX;
        const diffY = this.currY - this.prevY;
        ctx.strokeRect(this.currX - diffX, this.currY - diffY, diffX,diffY);
        clearTimeout(this.timerMarchingAnts)
        this.timerMarchingAnts = setTimeout(()=> {
            offsetDash++;
            this.draw(offsetDash);
        }, 100);
    }
    findxy(res, e) {
        if(this.props.changeMouseType.mouseType != "select") return;
        if (res == 'down') {
            this.currX = e.clientX - this.canvas.offsetLeft;
            this.currY = e.clientY - this.canvas.offsetTop;
            this.prevX = this.currX;
            this.prevY = this.currY;
            this.flag = true;
        }
        if (res == 'up' || res == "out") {
            this.flag = false;
        }
        
        if (res == 'move') {
            if (this.flag) {
                this.currX = e.clientX - this.canvas.offsetLeft;
                this.currY = e.clientY - this.canvas.offsetTop;
                this.draw(0);
            }
        }
    }
    render() {
        return <canvas className='canvaslayer'
            width={this.props.width}
            height={this.props.height}
            style={{width: this.props.width, height: this.props.height}}
            onMouseMove={(e)=> this.findxy('move', e)}
            onMouseDown={(e)=> this.findxy('down', e)}
            onMouseUp={(e)=> this.findxy('up', e)}
            onMouseOut={(e)=> this.findxy('out', e)}
            ref={(canvas)=>this.canvas = canvas} />
    }
}

const mapStateToProps = state => state;

export default connect(mapStateToProps)(SelectTool);