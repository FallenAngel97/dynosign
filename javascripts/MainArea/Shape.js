export class Shape {
  constructor (x, y, w, h, fill) {
    this.x = x || 0;
    this.y = y || 0;
    this.w = w || 0;
    this.h = h || 0;
    this.fill = fill || '#000';
    this.shapeType = 'rectangle';
  }
  setShape (shape) {
    this.shapeType = shape;
  }
  contains (clientX, clientY) {
    return (this.x <= clientX && this.x + this.w >= clientX) &&
      (this.y <= clientY && this.y + this.h >= clientY);
  }
  draw (ctx) {
    ctx.beginPath();
    switch (this.shapeType) {
      case 'rectangle':
        ctx.fillStyle = this.fill;
        ctx.fillRect(this.x, this.y, this.w, this.h);
        ctx.closePath();
        break;
      case 'circle':
        ctx.fillStyle = this.fill;
        ctx.arc(this.x, this.y, this.w, 0, 2 * Math.PI)
        ctx.fill()
        break;
    }
  }
}
