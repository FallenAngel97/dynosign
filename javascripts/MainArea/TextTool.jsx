import React from 'react';
import { connect } from 'react-redux';
import './TextTool.scss';
import PropTypes from 'prop-types';

export class TextTool extends React.Component {
  constructor (props) {
    super(props);
    this.addText = this.addText.bind(this);
  }
  addText (ev) {
    const input = document.getElementById('text_tool_input');
    if (input.contains(ev.target)) return;
    const x = ev.clientX - this.textWrapper.offsetLeft;
    const y = ev.clientY - this.textWrapper.offsetTop;
    input.style.cssText = `display: block; left: ${x}px; top: ${y}px`;
    input.focus()
  }
  componentWillUnmount () {
    // Save text logic
  }
  render () {
    return (
      <div ref={textWrapper => { this.textWrapper = textWrapper }} onClick={this.addText} id='text_tool_drawing_area'>
        <input style={{ fontFamily: this.props.changeFont.font.label }} id='text_tool_input' type='text' />
      </div>
    )
  };
}

TextTool.propTypes = {
  changeFont: PropTypes.object
}

const mapStateToProps = state => state;

export default connect(mapStateToProps)(TextTool);
