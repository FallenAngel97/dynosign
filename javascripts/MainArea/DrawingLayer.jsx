/* eslint-disable curly */
import React from 'react';
import { connect } from 'react-redux';
import { addLine } from '../actions';
import PropTypes from 'prop-types';
import { Shape } from './Shape';

export class DrawingLayer extends React.Component {
  constructor (props) {
    super(props);
    this.prevX = 0;
    this.currX = 0;
    this.prevX = 0;
    this.prevY = 0;
    this.dotFlag = false;
    this.flag = false;
    this.findxy = this.findxy.bind(this);
    this.draw = this.draw.bind(this);
    this.pasteToCanvas = this.pasteToCanvas.bind(this);
    this.ctrlPressed = false;
    this.dotWasDrawn = false;
    this.moveShape = this.moveShape.bind(this);
    this.shapes = [];
  }
  draw () {
    var ctx = this.canvas.getContext('2d');
    ctx.beginPath();
    ctx.restore();
    const color = this.props.changeColor.color;
    switch (this.props.changeMouseType.mouseType) {
      case 'draw':
        if (this.ctrlPressed) {
          ctx.moveTo(this.prevX, this.prevY);
          ctx.lineTo(this.prevX, this.currY);
        } else {
          ctx.moveTo(this.prevX, this.prevY);
          ctx.lineTo(this.currX, this.currY);
        }
        var y = 2;
        ctx.strokeStyle = `rgba(${color.r},${color.g},${color.b},${color.a})`;
        ctx.lineWidth = y;
        ctx.stroke();
        break;
      case 'rectangle':
        ctx.clearRect(0, 0, this.canvas.width, this.canvas.height);
        this.shapes.map((shape) => {
          ctx.fillStyle = shape.fill;
          ctx.fillRect(shape.x, shape.y, shape.w, shape.h);
        });
        ctx.fillStyle = `rgba(${color.r},${color.g},${color.b},${color.a})`;
        let diffX = this.currX - this.prevX;
        let diffY = this.currY - this.prevY;
        ctx.fillRect(this.currX - diffX, this.currY - diffY, diffX, diffY)
        break;
      case 'circle':
        ctx.clearRect(0, 0, this.canvas.width, this.canvas.height);
        ctx.fillStyle = `rgba(${color.r},${color.g},${color.b},${color.a})`;
        diffX = this.currX - this.prevX;
        diffY = this.currY - this.prevY;
        ctx.arc(this.currX - diffX, this.currY - diffY, Math.abs(diffX), 0, 2 * Math.PI)
        ctx.fill()
        break;
      case 'default':
        ctx.clearRect(0, 0, this.canvas.width, this.canvas.height);
        this.shapes.map((shape) => {
          ctx.fillStyle = shape.fill;
          ctx.fillRect(shape.x, shape.y, shape.w, shape.h);
        });
        break;
    }
    ctx.closePath();
  }
  moveShape (res, e) {
    if (res === 'down') {
      this.prevX = this.currX;
      this.prevY = this.currY;

      this.currX = e.clientX - this.canvas.offsetLeft;
      this.currY = e.clientY - this.canvas.offsetTop;
      this.flag = true;
    }
    if (res === 'up') {
      this.flag = false;
    }
    if (res === 'move') {
      if (this.flag) {
        this.prevX = this.currX;
        this.prevY = this.currY;
        this.currX = e.clientX - this.canvas.offsetLeft;
        this.currY = e.clientY - this.canvas.offsetTop;
        const k = this.shapes.length;
        for (let i = k - 1; i >= 0; i--) {
          if (this.shapes[i].contains(this.currX, this.currY)) {
            this.shapes[i].x = this.shapes[i].x + (this.currX - this.prevX);
            this.shapes[i].y = this.shapes[i].y + (this.currY - this.prevY);
          }
        }
        this.draw();
      }
    }
  }
  pasteToCanvas (e) {
    var items = e.clipboardData.items;
    console.log(items)
  }
  findxy (res, e) {
    if (this.props.changeMouseType.mouseType === 'default') {
      this.moveShape(res, e);
      return;
    }
    if (!/draw|circle|rectangle/.test(this.props.changeMouseType.mouseType)) return;
    this.ctrlPressed = false;
    if (e.ctrlKey) {
      this.ctrlPressed = true;
    }
    if (res === 'down') {
      this.prevX = this.currX;
      this.prevY = this.currY;

      this.currX = e.clientX - this.canvas.offsetLeft;
      this.currY = e.clientY - this.canvas.offsetTop;

      if (this.props.changeMouseType.mouseType === 'rectangle' || this.props.changeMouseType.mouseType === 'circle' || this.ctrlPressed) {
        this.prevX = this.currX;
        this.prevY = this.currY;
      }

      this.flag = true;
      if (this.props.changeMouseType.mouseType === 'draw') {
        this.dotWasDrawn = true;
        this.dotFlag = true;
      }
      if (this.dotFlag) {
        var ctx = this.canvas.getContext('2d');
        ctx.beginPath();
        const color = this.props.changeColor.color;
        if (e.shiftKey) {
          ctx.strokeStyle = `rgba(${color.r},${color.g},${color.b},${color.a})`;
          ctx.moveTo(this.prevX, this.prevY);
          ctx.lineTo(this.currX, this.currY);
          ctx.stroke();
        } else {
          ctx.fillStyle = `rgba(${color.r},${color.g},${color.b},${color.a})`;
          ctx.fillRect(this.currX, this.currY, 2, 2);
          ctx.closePath();
        }
        this.dotFlag = false;
      }
    }
    if (res === 'up' || res === 'out') {
      this.flag = false;
    }

    if (res === 'up') {
      this.props.addLine(this.canvas, this.props.changeActiveLayer.layerNumber);
      if (this.props.changeMouseType.mouseType === 'rectangle') {
        const diffX = this.currX - this.prevX;
        const diffY = this.currY - this.prevY;
        const ctx = this.canvas.getContext('2d');
        this.shapes.push(new Shape(this.currX - diffX, this.currY - diffY, diffX, diffY, ctx.fillStyle));
      }
    }

    if (res === 'move') {
      if (this.flag) {
        this.dotWasDrawn = false;
        if (this.props.changeMouseType.mouseType === 'draw') {
          if (!this.ctrlPressed)
            this.prevX = this.currX;
          this.prevY = this.currY;
        }
        this.currX = e.clientX - this.canvas.offsetLeft;
        this.currY = e.clientY - this.canvas.offsetTop;
        this.draw();
      }
    }
  }
  render () {
    return (
      <canvas className='canvaslayer'
        width={this.props.width}
        height={this.props.height}
        style={{ width: this.props.width, height: this.props.height }}
        onMouseMove={(e) => this.findxy('move', e)}
        onMouseDown={(e) => this.findxy('down', e)}
        onMouseUp={(e) => this.findxy('up', e)}
        onMouseOut={(e) => this.findxy('out', e)}
        onPaste={(e) => this.pasteToCanvas(e)}
        ref={(canvas) => { this.canvas = canvas }} />
    )
  }
}

DrawingLayer.propTypes = {
  changeColor: PropTypes.object,
  changeMouseType: PropTypes.object,
  height: PropTypes.number,
  width: PropTypes.number,
  addLine: PropTypes.func,
  changeActiveLayer: PropTypes.object
}

const mapStateToProps = state => state;

export const mapDispatchToProps = dispatch => {
  return {
    addLine: (elem, layerNumber) => dispatch(addLine(elem, layerNumber))
  }
}

export default connect(mapStateToProps, mapDispatchToProps, null, { forwardRef: true })(DrawingLayer);
