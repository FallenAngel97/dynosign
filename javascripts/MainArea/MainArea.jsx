import React from "react";
import "./MainArea.scss";
import {connect} from "react-redux";
import {addLine, redoLine} from "../actions";

class MainArea extends React.Component {
    constructor(props) {
        super(props);

        this.prevX = 0;
        this.currX = 0;
        this.prevX = 0;
        this.prevY = 0;
        this.dot_flag = false;
        this.flag = false;

        this.findxy = this.findxy.bind(this);
        this.draw = this.draw.bind(this);

        this.layers = [];
    }
    componentDidMount() {
        window.addEventListener('keydown', (e) => {
            if(e.keyCode == 90 && e.ctrlKey) {
                this.props.redoLine(this.props.changeActiveLayer.layerNumber);
            }
        }, false)

        this.layers.map((ref)=>{
            ref.style.width = this.layerContainer.offsetWidth;
            ref.style.height = this.layerContainer.offsetHeight;
            ref.height = this.layerContainer.offsetHeight;
            ref.width = this.layerContainer.offsetWidth;
            ref.addEventListener("mousemove",  (e) => {
                if(this.props.changeMouseType.mouseType == 'draw' || this.props.changeMouseType.mouseType == 'select')
                    this.findxy('move', e)
            }, false);
            ref.addEventListener("mousedown", (e) => {
                if(this.props.changeMouseType.mouseType == 'draw' || this.props.changeMouseType.mouseType == 'select')
                    this.findxy('down', e)
            }, false);
            ref.addEventListener("mouseup", (e) => {
                if(this.props.changeMouseType.mouseType == 'draw' || this.props.changeMouseType.mouseType == 'select')
                    this.findxy('up', e)
            }, false);
            ref.addEventListener("mouseout", (e) => {
                if(this.props.changeMouseType.mouseType == 'draw' || this.props.changeMouseType.mouseType == 'select')
                    this.findxy('out', e)
            }, false);
        });
    }
    draw() {
        const layerIndex = this.props.changeActiveLayer.layerNumber;

        var ctx = this.layers[layerIndex].getContext('2d');
        ctx.beginPath();
        if(this.props.changeMouseType.mouseType == 'draw') {
            ctx.restore();
            ctx.moveTo(this.prevX, this.prevY);
            ctx.lineTo(this.currX, this.currY);
            var x = "black",
            y = 2;
            ctx.strokeStyle = x;
            ctx.lineWidth = y;
            ctx.stroke();
            ctx.closePath();
        }
        else {
            ctx.clearRect(0, 0, this.layers[layerIndex].width, this.layers[layerIndex].height);
            ctx.setLineDash([6])
            const diffX = this.currX - this.prevX;
            const diffY = this.currY - this.prevY;
            ctx.strokeRect(this.currX - diffX, this.currY - diffY, diffX,diffY);
        }
    }
    findxy(res, e) {
        const layerIndex = this.props.changeActiveLayer.layerNumber;

        if (res == 'down') {
            this.prevX = this.currX;
            this.prevY = this.currY;

            this.currX = e.clientX - this.layers[layerIndex].offsetLeft;
            this.currY = e.clientY - this.layers[layerIndex].offsetTop;
            if(this.props.changeMouseType.mouseType == 'select') {
                this.prevX = this.currX;
                this.prevY = this.currY;
            }
            this.flag = true;
            this.dot_flag = true;
            if (this.dot_flag) {
                var ctx = this.layers[layerIndex].getContext('2d');
                ctx.beginPath();
                ctx.fillStyle = 'black';
                ctx.fillRect(this.currX, this.currY, 2, 2);
                ctx.closePath();
                this.dot_flag = false;
            }
        }
        if (res == 'up' || res == "out") {
            this.flag = false;
        }
        
        if(res=="up"){
            if(this.props.changeMouseType.mouseType == 'draw')
                this.props.addLine(this.layers[layerIndex], this.props.changeActiveLayer.layerNumber);
            else {

            }
        }
        
        if (res == 'move') {
            if (this.flag) {
                if(this.props.changeMouseType.mouseType == 'draw') {
                    this.prevX = this.currX;
                    this.prevY = this.currY;
                }
                this.currX = e.clientX - this.layers[layerIndex].offsetLeft;
                this.currY = e.clientY - this.layers[layerIndex].offsetTop;
                this.draw();
            }
        }
    }
    componentDidUpdate(prevProps) {
        this.layers.map((ref)=>{
            if(ref && ref.getAttribute("height") == undefined) {
                ref.style.width = this.layerContainer.offsetWidth;
                ref.style.height = this.layerContainer.offsetHeight;
                ref.height = this.layerContainer.offsetHeight;
                ref.width = this.layerContainer.offsetWidth;
                ref.addEventListener("mousemove",  (e) => {
                    if(this.props.changeMouseType.mouseType == 'draw' || this.props.changeMouseType.mouseType == 'select')
                        this.findxy('move', e)
                }, false);
                ref.addEventListener("mousedown", (e) => {
                    if(this.props.changeMouseType.mouseType == 'draw' || this.props.changeMouseType.mouseType == 'select')
                        this.findxy('down', e)
                }, false);
                ref.addEventListener("mouseup", (e) => {
                    if(this.props.changeMouseType.mouseType == 'draw' || this.props.changeMouseType.mouseType == 'select')
                        this.findxy('up', e)
                }, false);
                ref.addEventListener("mouseout", (e) => {
                    if(this.props.changeMouseType.mouseType == 'draw' || this.props.changeMouseType.mouseType == 'select')
                        this.findxy('out', e)
                }, false);
            }
        });
        if(this.props.layersCRUD!=prevProps.layersCRUD) {
            const layerIndex = this.props.changeActiveLayer.layerNumber;
            let layerImg = new Image();
            const linesArray = this.props.layersCRUD[layerIndex].linesArray;

            if(linesArray.length == 0) {
                let ctx = this.layers[layerIndex].getContext('2d');
                ctx.clearRect(0, 0, this.layers[layerIndex].width, this.layers[layerIndex].height);
            } else {
                if(this.props.layersCRUD[layerIndex].hidden == true) return;
                layerImg.src = linesArray[linesArray.length-1];
                layerImg.onload = () =>{ 
                    let ctx = this.layers[layerIndex].getContext('2d');
                    ctx.clearRect(0, 0, this.layers[layerIndex].width, this.layers[layerIndex].height);
                    ctx.drawImage(layerImg, 0, 0); 
                }
            }

            this.props.layersCRUD.map((layer, index)=> {
                if(layer.hidden == true) {
                    let ctx = this.layers[index].getContext('2d');
                    ctx.clearRect(0, 0, this.layers[index].width, this.layers[index].height);
                }
            });

        }
    }
    render() {
        let iconType = 'default';
        switch(this.props.changeMouseType.mouseType) {
            case "select":
                iconType = 'crosshair';
                break;
            case "text":
                iconType = 'text';
                break;
            case "draw":
                iconType = 'crosshair';
                break;
            case "circle":
                iconType = 'crosshair';
                break;
        }
        return (
            <div id='mainArea'>
                <button className="drawAnimateSelector">Draw</button>
                <button className="drawAnimateSelector">Animate</button>
                <div ref={layerContainer => this.layerContainer = layerContainer} style={{cursor: iconType}} id='drawingArea' >
                    {this.props.layersCRUD.map((layer, index)=> {
                        return <canvas className='canvaslayer' key={index} ref={(_ref)=>this.layers[index] = _ref} />
                    })}
                </div>
            </div>
        )
    }
}

const mapStateToProps = state => state;

const mapDispatchToProps = dispatch => {
    return {
        addLine: (elem, layerNumber) => dispatch(addLine(elem, layerNumber)),
        redoLine: (layerNumber) => dispatch(redoLine(layerNumber))
    }
}

export default connect(mapStateToProps, mapDispatchToProps)(MainArea);