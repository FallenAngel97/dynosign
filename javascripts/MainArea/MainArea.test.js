import React from "react"
import { MainArea, mapDispatchToProps } from "./MainArea";
import { configure, mount, shallow} from "enzyme"
import Adapter from 'enzyme-adapter-react-16';
configure({ adapter: new Adapter() })

describe("MainArea tests", () => {
    test("MainArea should render properly", () => {
        let mainArea = mount(<MainArea layersCRUD={[]} changeMouseType={{mouseType: 'default'}} />);
        expect(mainArea).toMatchSnapshot();
        expect(mainArea.instance().layers).toBeTruthy();
    })
})