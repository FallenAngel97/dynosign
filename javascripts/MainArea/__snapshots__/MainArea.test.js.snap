// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MainArea tests MainArea should render properly 1`] = `ShallowWrapper {}`;
